internal/eventstream/client.go

```diff
@@ -0,0 +1,391 @@
+	db *icingadb.DB,
+	runtimeConfig *config.RuntimeConfig,
+	conf *config.ConfigFile,
+) ([]*Client, error) {
+	clients := make([]*Client, 0, len(conf.Icinga2Apis))
+
+	for _, icinga2Api := range conf.Icinga2Apis {
+		logger := logs.GetChildLogger(fmt.Sprintf("eventstream-%d", icinga2Api.NotificationsEventSourceId))
+
+		client := &Client{
+			ApiHost:          icinga2Api.Host,
+			ApiBasicAuthUser: icinga2Api.AuthUser,
+			ApiBasicAuthPass: icinga2Api.AuthPass,
+			ApiHttpTransport: http.Transport{
+				// Hardened TLS config adjusted to Icinga 2's configuration:
+				// - https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#objecttype-apilistener
+				// - https://icinga.com/docs/icinga-2/latest/doc/12-icinga2-api/#security
+				// - https://ssl-config.mozilla.org/#server=go&config=intermediate
+				TLSClientConfig: &tls.Config{
+					MinVersion: tls.VersionTLS12,
+					CipherSuites: []uint16{
+						tls.TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+						tls.TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+						tls.TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+						tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+						tls.TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305,
+						tls.TLS_ECDHE_RSA_WITH_CHACHA20_POLY1305,
+					},
+				},
+			},
+
+			IcingaNotificationsEventSourceId: icinga2Api.NotificationsEventSourceId,
+			IcingaWebRoot:                    conf.Icingaweb2URL,
+
+			CallbackFn: makeProcessEvent(ctx, db, logger, logs, runtimeConfig, conf),
+			Ctx:        ctx,
+			Logger:     logger,
+		}
+
+		if icinga2Api.IcingaCaFile != "" {
+			caData, err := os.ReadFile(icinga2Api.IcingaCaFile)
+			if err != nil {
+				return nil, fmt.Errorf("cannot read CA file %q for Event Stream ID %d, %w",
+					icinga2Api.IcingaCaFile, icinga2Api.NotificationsEventSourceId, err)
+			}
+
+			certPool := x509.NewCertPool()
+			if !certPool.AppendCertsFromPEM(caData) {
+				return nil, fmt.Errorf("cannot add custom CA file to CA pool for Event Stream ID %d, %w",
+					icinga2Api.NotificationsEventSourceId, err)
+			}
+
+			client.ApiHttpTransport.TLSClientConfig.RootCAs = certPool
+		}
+
+		if icinga2Api.InsecureTls {
+			client.ApiHttpTransport.TLSClientConfig.InsecureSkipVerify = true
+		}
+
+		clients = append(clients, client)
+	}
+	return clients, nil
+}
+
+// buildCommonEvent creates an event.Event based on Host and (optional) Service attributes to be specified later.
+//
+// The new Event's Time will be the current timestamp.
+//
+// The following fields will NOT be populated and might be altered later:
+//   - Type
+//   - Severity
+//   - Username
+//   - Message
+//   - ID
+func (client *Client) buildCommonEvent(host, service string) (*event.Event, error) {
+	var (
+		eventName      string
+		eventUrlSuffix string
+		eventTags      map[string]string
+		eventExtraTags = make(map[string]string)
+	)
+
+	if service != "" {
+		eventName = host + "!" + service
+		eventUrlSuffix = "/icingadb/service?name=" + url.PathEscape(service) + "&host.name=" + url.PathEscape(host)
```

Comment:
Shouldn't this use `url.QueryEscape()`?

Reply:
Unfortunately `url.QueryEscape(" "))` results in `+` instead of `%20` which breaks the URL for Icinga Web.

Initially I tried using `url.QueryEscape` but had to realize that it doesn't work for this use case.

Reply:
`url.PathEscape()` misses other characters like `&`: https://go.dev/play/p/z4lO-ZYlSwp

From my quick look through the `net/url` code, looks like there's no function that does exactly what we would need here then. Soo... `QueryEscape()` and replace `+` with `%20`? :see_no_evil: (And maybe open an issue in Icinga Web for that.)

Reply:
Needs more SSE2: https://github.com/php/php-src/blob/php-8.2.12/ext/standard/url.c#L472